Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,143 @@
+            self.handleAction(action, context: context)
+        } catch {
+            context.fireErrorCaught(error)
+            context.close(mode: .all, promise: nil)
+        }
+    }
+    
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        guard self.state.proxyEstablished else {
+            self.bufferedWrites.append((data, promise))
+            return
+        }
+        self.writeBufferedData(context: context)
+        context.write(data, promise: nil)
+    }
+    
+    public func writeBufferedData(context: ChannelHandlerContext) {
+        while let (data, promise) = self.bufferedWrites.first {
+            context.write(data, promise: promise)
+            self.bufferedWrites.removeFirst()
+        }
+    }
+    
+}
+
+extension SOCKSClientHandler {
+    
+    func beginHandshake(context: ChannelHandlerContext) {
+        guard self.state.shouldBeginHandshake else {
+            return
+        }
+        self.handleAction(self.state.connectionEstablished(), context: context)
+    }
+    
+    func handleAction(_ action: ClientAction, context: ChannelHandlerContext) {
+        switch action {
+        case .waitForMoreData:
+            break // do nothing, we've already buffered the data
+        case .sendGreeting:
+            self.handleActionSendClientGreeting(context: context)
+        case .sendRequest:
+            self.handleActionSendRequest(context: context)
+        case .proxyEstablished:
+            self.handleActionProxyEstablished(context: context)
+        case .sendData(let data):
+            context.writeAndFlush(self.wrapOutboundOut(data), promise: nil)
+        }
+    }
+    
+    func handleActionSendClientGreeting(context: ChannelHandlerContext) {
+        let greeting = ClientGreeting(methods: [.noneRequired]) // no authentication currently supported
+        let capacity = 1 + 1 + 1 // [version, #methods, methods...]
+        var buffer = context.channel.allocator.buffer(capacity: capacity)
+        buffer.writeClientGreeting(greeting)
+        self.state.sendClientGreeting(greeting)
+        context.writeAndFlush(self.wrapOutboundOut(buffer), promise: nil)
+    }
+    
+    func handleActionProxyEstablished(context: ChannelHandlerContext) {
+        // for some reason we have extra bytes
+        // so let's send them down the pipe
+        if self.buffered.readableBytes > 0 {
+            let data = self.wrapInboundOut(self.buffered)
+            context.fireChannelRead(data)
+        }
+        
+        // If we have any buffered writes then now
+        // we can send them.
+        self.writeBufferedData(context: context)
+    }
+    
+    func handleActionSendRequest(context: ChannelHandlerContext) {
+        let request = ClientRequest(command: .connect, addressType: self.targetAddress)
+        self.state.sendClientRequest(request)
+        
+        // the client request is always 5 bytes + the address info
```

Comment:
This comment seems wrong: isn't it 6 bytes + address info?

Reply:
Yes, maths